File: ios/Ninehoto/Utilities/DateFormatters.swift
import Foundation

final class DateFormatters {
    static let shared = DateFormatters()
    
    private let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        return formatter
    }()
    
    private let shortDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateStyle = .short
        return formatter
    }()
    
    private let relativeFormatter: RelativeDateTimeFormatter = {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return formatter
    }()
    
    private let iso8601Formatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()
    
    private let timeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.timeStyle = .short
        return formatter
    }()
    
    private init() {}
    
    func format(_ date: Date) -> String {
        dateFormatter.string(from: date)
    }
    
    func formatShort(_ date: Date) -> String {
        shortDateFormatter.string(from: date)
    }
    
    func formatRelative(_ date: Date) -> String {
        relativeFormatter.localizedString(for: date, relativeTo: Date())
    }
    
    func formatISO8601(_ date: Date) -> String {
        iso8601Formatter.string(from: date)
    }
    
    func formatTime(_ date: Date) -> String {
        timeFormatter.string(from: date)
    }
    
    func parse(_ string: String) -> Date? {
        dateFormatter.date(from: string)
    }
    
    func parseISO8601(_ string: String) -> Date? {
        iso8601Formatter.date(from: string)
    }
    
    func timeAgo(_ date: Date) -> String {
        let calendar = Calendar.current
        let now = Date()
        let components = calendar.dateComponents([.minute, .hour, .day, .weekOfYear, .month, .year], from: date, to: now)
        
        if let year = components.year, year > 0 {
            return year == 1 ? "1 year ago" : "\(year) years ago"
        }
        if let month = components.month, month > 0 {
            return month == 1 ? "1 month ago" : "\(month) months ago"
        }
        if let week = components.weekOfYear, week > 0 {
            return week == 1 ? "1 week ago" : "\(week) weeks ago"
        }
        if let day = components.day, day > 0 {
            return day == 1 ? "Yesterday" : "\(day) days ago"
        }
        if let hour = components.hour, hour > 0 {
            return hour == 1 ? "1 hour ago" : "\(hour) hours ago"
        }
        if let minute = components.minute, minute > 0 {
            return minute == 1 ? "1 minute ago" : "\(minute) minutes ago"
        }
        
        return "Just now"
    }
}

extension Date {
    var formatted: String { DateFormatters.shared.format(self) }
    var formattedShort: String { DateFormatters.shared.formatShort(self) }
    var formattedRelative: String { DateFormatters.shared.formatRelative(self) }
    var formattedISO8601: String { DateFormatters.shared.formatISO8601(self) }
    var formattedTime: String { DateFormatters.shared.formatTime(self) }
    var timeAgo: String { DateFormatters.shared.timeAgo(self) }
}

File: android/Ninehoto.Android/Utilities/DateFormatters.cs
using System;
using System.Globalization;

namespace Ninehoto.Android.Utilities
{
    public class DateFormatters
    {
        private static DateFormatters? _instance;
        public static DateFormatters Instance => _instance ??= new DateFormatters();

        private readonly Java.Text.DateFormat _dateFormatter;
        private readonly Java.Text.DateFormat _shortDateFormatter;
        private readonly Java.Text.DateFormat _timeFormatter;

        private DateFormatters()
        {
            _dateFormatter = Java.Text.DateFormat.GetDateTimeInstance(Java.Text.DateFormat.Medium, Java.Text.DateFormat.Short, CultureInfo.CurrentCulture);
            _shortDateFormatter = Java.Text.DateFormat.GetDateInstance(Java.Text.DateFormat.Short, CultureInfo.CurrentCulture);
            _timeFormatter = Java.Text.DateFormat.GetTimeInstance(Java.Text.DateFormat.Short, CultureInfo.CurrentCulture);
        }

        public string Format(DateTime date)
        {
            return _dateFormatter.Format(date);
        }

        public string FormatShort(DateTime date)
        {
            return _shortDateFormatter.Format(date);
        }

        public string FormatTime(DateTime date)
        {
            return _timeFormatter.Format(date);
        }

        public string TimeAgo(DateTime date)
        {
            var diff = DateTime.Now - date;

            if (diff.TotalDays >= 365)
            {
                var years = (int)(diff.TotalDays / 365);
                return years == 1 ? "1 year ago" : $"{years} years ago";
            }
            if (diff.TotalDays >= 30)
            {
                var months = (int)(diff.TotalDays / 30);
                return months == 1 ? "1 month ago" : $"{months} months ago";
            }
            if (diff.TotalDays >= 7)
            {
                var weeks = (int)(diff.TotalDays / 7);
                return weeks == 1 ? "1 week ago" : $"{weeks} weeks ago";
            }
            if (diff.TotalDays >= 1)
            {
                var days = (int)diff.TotalDays;
                return days == 1 ? "Yesterday" : $"{days} days ago";
            }
            if (diff.TotalHours >= 1)
            {
                var hours = (int)diff.TotalHours;
                return hours == 1 ? "1 hour ago" : $"{hours} hours ago";
            }
            if (diff.TotalMinutes >= 1)
            {
                var minutes = (int)diff.TotalMinutes;
                return minutes == 1 ? "1 minute ago" : $"{minutes} minutes ago";
            }

            return "Just now";
        }

        public DateTime? Parse(string dateString)
        {
            try
            {
                return _dateFormatter.Parse(dateString);
            }
            catch
            {
                return null;
            }
        }

        public DateTime? ParseISO8601(string dateString)
        {
            try
            {
                return System.DateTime.Parse(dateString, null, System.Globalization.DateTimeStyles.RoundtripKind);
            }
            catch
            {
                return null;
            }
        }

        public string FormatISO8601(DateTime date)
        {
            return date.ToString("o", CultureInfo.InvariantCulture);
        }
    }

    public static class DateExtensions
    {
        public static string Formatted(this DateTime date) => DateFormatters.Instance.Format(date);
        public static string FormattedShort(this DateTime date) => DateFormatters.Instance.FormatShort(date);
        public static string FormattedTime(this DateTime date) => DateFormatters.Instance.FormatTime(date);
        public static string TimeAgo(this DateTime date) => DateFormatters.Instance.TimeAgo(date);
    }
}